omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+        # just in case someone wants to give tfs files (hidden feature)
+        df = tfs.read_tfs(file_path)
+    else:
+        # parse csv file (the official way)
+        dtypes = {"madx": str, "lsa": str, "scaling": float, "test": float}
+        converters = {'madx': str.strip, 'lsa': str.strip}  # strip whitespaces
+        df = pd.read_csv(file_path, comment="#", names=dtypes.keys(), dtype=dtypes, converters=converters)
+    return _dataframe_to_knobsdict(df)
+
+
+def _dataframe_to_knobsdict(df: pd.DataFrame) -> KnobsDict:
+    """ Converts a DataFrame into the required Dictionary structure.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing at least the columns
+                           'lsa', 'madx', 'scaling' (upper or lowercase)
+
+    Returns:
+        Dictionary with LSA names (but with colon instead of /) as
+        keys and KnobEntries (without values) as value.
+
+    """
+    df.columns = df.columns.astype(str).str.lower()
+    df = df[['lsa', 'madx', 'scaling']].set_index("lsa", drop=False)
+    return {
+        lsa2name(r[0]): KnobEntry(**r[1].to_dict()) for r in df.iterrows()
+    }
+
+
+def _parse_knobs_defintions(knobs_def_input: Optional[Union[Path, str, pd.DataFrame]]) -> KnobsDict:
+    """ Parse the given knob-definitions either from a csv-file or from a DataFrame. """
+    if isinstance(knobs_def_input, pd.DataFrame):
+        return _dataframe_to_knobsdict(knobs_def_input)
+
+    # input points to a file or is None
+    knobs_def_file = _get_knobs_def_file(knobs_def_input)
+    return _load_knobs_dict(knobs_def_file)
+
+
+# Time Tools -------------------------------------------------------------------
+
+def _parse_time(time: str, timedelta: str = None) -> datetime:
+    """ Parse time from given time-input. """
+    t = _parse_time_from_str(time)
+    if timedelta:
+        t = _add_time_delta(t, timedelta)
+    return t
+
+
+def _parse_time_from_str(time_str: str) -> datetime:
+    """ Parse time from given string. """
+    # Now? ---
+    if time_str.lower() == "now":
+        return datetime.now()
+
+    # ISOFormat? ---
+    try:
+        return datetime.fromisoformat(time_str)
+    except (TypeError, ValueError):
+        pass
+
+    # Timestamp? ---
+    try:
+        return datetime.fromtimestamp(int(time_str))
+    except (TypeError, ValueError):
+        pass
+
+    raise ValueError(f"Couldn't read datetime '{time_str}'")
+
+
+def _add_time_delta(time: datetime, delta_str: str) -> datetime:
+    """ Parse delta-string and add time-delta to time. """
+    sign = -1 if delta_str[0] in MINUS_CHARS else 1
+    all_deltas = re.findall(r"(\d+)(\w)", delta_str)
+
+    # following ISO-8601 for time durations
+    char_map = dict(
+        s='seconds', m='minutes', h='hours',
+        d='days', w='weeks', M='months', Y="years",
+    )
+
+    # add all deltas
+    time_parts = {char_map[delta[1]]: sign * int(delta[0]) for delta in all_deltas}
+    time = time + relativedelta(**time_parts)
+
```

Comment:
this is a bit hard to understand. this can arguably be left as-is, but it _should_ be either less confusing or have some comments explaining what is going on.

Reply:
I agree, it took me some time to get it.
Once you know what it does it's fine, but a comment would make it obvious at the first sight

Reply:
Agreed, the comprehension is succint but the comment line above could detail what each of the indices used represent. Don't want to scroll the file to look for what `delta[1]` is etc.